llvm/lib/Analysis/VectorUtils.cpp

```diff
@@ -413,6 +413,36 @@ bool llvm::getShuffleDemandedElts(int SrcWidth, ArrayRef<int> Mask,
   return true;
 }
 
+bool llvm::isMaskedSlidePair(ArrayRef<int> Mask, int NumElts,
+                             std::array<std::pair<int, int>, 2> &SrcInfo) {
+  const int SignalValue = NumElts * 2;
+  SrcInfo[0] = {-1, SignalValue};
+  SrcInfo[1] = {-1, SignalValue};
```

Comment:
For length changing shuffles is it possible for a mask index to be equal to the signal value?

Reply:
Not to my knowledge.  A valid shuffle should have a maximum index of 2 x the number of elements in the result - 1 at most.   This is asserted in variously places in code - the most important of which is ShuffleVectorInst::isValidOperands.

Reply:
Ok, the code is actually correct, but only somewhat accidentally.  The getShuffleCost API is a bit of a mess.  The type parameter is sometimes the source type, and sometimes the destination type.  For the PermuteSingleSrc and PermuteTwoSrc forms, at least the getInstructionCost code normalizes to the wider type, and passes in that one.  (i.e. from the costing perspective, there is no length changing shuffle.)

I have not audited other callers of this API, and make no claims about whether they handle this correctly.  However, given the above code, if they don't, it would appear to be a bug in the caller.  

I hate this API for the record.  It's full of these special cases and it makes working on it really hard to do without introducing bugs.

Reply:
Agreed, this API is mind bending to say the least. I just stuck in an assert to see if the users are normalizing the mask to the widest type:

```
  assert(!(Kind == TTI::SK_PermuteSingleSrc || Kind == TTI::SK_PermuteTwoSrc) ||
         all_of(Mask, [&Mask](int x) {
           return x == -1 || x < (int)Mask.size() * 2;
         }));
```

Annoyingly, `ninja check-llvm-analysis-costmodel-riscv check-llvm-transforms-slpvectorizer` passes, but `ninja check-llvm-transforms-loopvectorize` triggers the assert. 

I would hope the loop vectorizer isn't emitting any masked slide shuffles though. Maybe the easiest thing to do for now is just to assert that all mask values are < Mask.size() * 2 in isMaskedSlidePair in case that ever changes?

Reply:
I went digging a bit further, and SLP is costing length changing shuffles in numerous places.  I decide to go with the easy fix here, and just add an explicit guard that avoids costing length changing shuffles with the new lowering.